compiler/src/dotty/tools/dotc/typer/Implicits.scala

```diff
@@ -928,9 +933,70 @@ trait Implicits:
           if normalImports.isEmpty then suggestedImports else normalImports
         end hiddenImplicitsAddendum
 
-        msg(userDefined.getOrElse(
-          em"no implicit argument of type $pt was found${location("for")}"))() ++
-        hiddenImplicitsAddendum
+
+        /** Extracting the message from a method parameter, e.g. in
+         *
+         *  trait Foo
+         *
+         *  def foo(implicit @annotation.implicitNotFound foo: Foo)
+         */
+        def userDefinedParamMessage = paramSymWithMethodTree.flatMap { (sym, tree) =>
+          userDefinedMsg(sym, defn.ImplicitNotFoundAnnot).map { rawMsg =>
+            val paramSyms = typeVariablesInScope(sym)
+            val paramTypeRefs = paramSyms.map(_.typeRef.typeConstructor)
+
+            val argTypes = tree match {
+              case TypeApply(Select(qual, _), targs) =>
+                val methodOwnerType = qual.tpe
+                val methodOwner = tree.symbol.owner
+                val methodTypeParams = tree.symbol.paramSymss.flatten.filter(_.isType)
+                val methodTypeArgs = targs.map(_.tpe)
+                paramTypeRefs.map(_.asSeenFrom(methodOwnerType, methodOwner).subst(methodTypeParams, methodTypeArgs))
+              case _ => paramTypeRefs
+            }
+
+            err.userDefinedErrorString(
+              rawMsg,
+              paramSyms.map(_.name.unexpandedName.toString),
+              argTypes
+            )
+          }
+        }
+
+        /** Extracting the message from a type, e.g. in
+         *
+         *  implicit @annotation.implicitNotFound
+         *  trait Foo
+         *
+         *  def foo(foo: Foo)
+         */
+        def userDefinedTypeMessage = userDefinedMsg(pt.typeSymbol, defn.ImplicitNotFoundAnnot).map { rawMsg =>
+          def extractTypeParams(tpe: Type): List[(String, Type)] =
```

Comment:
I realize this was broken before so doesn't really need to be addressed in this PR, but just looking at the type parameters of the type isn't enough, e.g.:
```scala
import scala.annotation.implicitNotFound

class Outer[A] {
  @implicitNotFound("A = ${A} -- B = ${B}")
  class Inner[B]
}

object Test {
  val x: Outer[Int] = new Outer[Int]
  implicitly[x.Inner[Double]]
}
```
I'd expect `A = Int -- B = Double` but instead I get `A =  -- B = Double`, Scala 2 has the same issue! I think a solution based on `asSeenFrom` like what is done in `userDefinedParamMessage` would make more sense, but I haven't tried it.